### ISSUE TITLE
Aircraft climbing on glideslope on short final

### ISSUE THREAD

**Issue by nalcodmnj** [openscope/openscope#1395]
Browser:
Chrome
Airport:
Boston Logan
Approximate play time before issue:

#### Issue Description
Flights end up increasing altitude after descending to about 1000 Ft while approaching runway 27. 
After clearing for ILS 27 flights end up way after runway 27.
![OBSERVE NKS 5577](https://user-images.githubusercontent.com/34276254/58760759-23760a00-855a-11e9-8e22-209e7f851fe2.png)
![NKS 5577 HAS STARTED DESCENDING](https://user-images.githubusercontent.com/34276254/58760769-4b656d80-855a-11e9-930f-f2f1f2809fcc.png)
![NKS 5577 RISING](https://user-images.githubusercontent.com/34276254/58760773-56200280-855a-11e9-9ef9-61702b7fbea3.png)
![KEPT ON RISING  UPTO 2500](https://user-images.githubusercontent.com/34276254/58760775-5b7d4d00-855a-11e9-87ad-fb6f536030f4.png)
![THEN ARRIVES](https://user-images.githubusercontent.com/34276254/58760777-5f10d400-855a-11e9-8c51-264fec9f8bb3.png)

#### Steps to reproduce
-

**Comment by erikquinn** [comment]
Is this actually happening to all airports, or is it just Boston? It's probably all if it's doing that... :grimacing:

**Comment by JPmAn24** [comment]
@erikquinn If you look at #1013, whenever an aircraft is cleared for runway 4 at KHOU, it does the same thing, yet if it is on 13L/R, it doesn’t do the bounce. Same for KBOS: Landing on 27 (which isn’t the best runway for wind) does the bounce, while landing on 22L/R doesn’t. I’ll do some investigations to see what would possibly be causing this.

**Comment by NathanDJohnson** [comment]
Same for runway 3 at ABQ.

**Comment by erikquinn** [comment]
A go-around message is NOT given. They simply climb after a certain point, then go back down again (and land after the airport, slowing and despawning as normal. Also some zig-zagging on the localizer within about 1nm final.

Could be something with the getOffset() logic? Seems like they must be trying to follow the glideslope and it's giving them bad information on the correct altitude to be targeting.... 🤔

**Comment by erikquinn** [comment]
Stems from `AircraftModel._calculateTargetedHeadingDuringLanding()`. Aircraft are going around because the allowable localizer deviation gets smaller as you get closer to the runway. For some reason, the headings being generated by this method are slightly wrong (far less than magnetic-variation-errors), resulting in a gentle drift off-course and subsequent go-around (to MCP altitude).

Both `aircraftPositionModel.bearingToPosition(runwayModel.positionModel)` and `bearingToPoint(aircraftPosition, runwayPosition)` yield similarly erroneous headings, as does simply flying `runwayModel.angle`.

If we fix this heading math, the aircraft will both stop making the goofy wiggle turns AND stop sending themselves around. More investigation needed.

**Comment by erikquinn** [comment]
Another possible cause could lie within the wind crab angle physics, which would explain why the aircraft are attempting to turn off centerline when they touch down. Usually, they slow down fast enough that you barely see it, but when I tell landed aircraft to slow to 20 knots instead of 100, they never despawn (due to not reaching 0kts), and we can see their flight path. Based on the screenshot, they clearly are not aiming for runway heading after landing runway 27... whether or not this is related to the problems brought up by this particular issue is unclear, but it's worth investigating as a possibility.

![image](https://user-images.githubusercontent.com/5103735/61554821-d8da1f80-aa2b-11e9-9ed6-ee0d216bd427.png)

**Comment by erikquinn** [comment]
Okay, definitely caused by aircraft not crabbing into the wind. On develop, here's what's happening:
- On the localizer, outside of a 4 mile final, aircraft heading is determined based on their offset from the localizer and remaining distance to the threshold. It is NOT based on the heading to the runway.
- On the localizer, within a 4 mile final, aircraft heading is based on the heading to the runway. This heading is where the aircraft is facing, and does not include any crabbing into the wind. Thus, in high crosswind situations, you can see the aircraft being blown off course, and taking a steeper and steeper angle to the runway since they never attempt to compensate for the fact that they're off centerline; they only continue aiming at the runway while the wind pushes them into a steeper and steeper approach. Eventually they reach the runway, and turn back to align with it.
- After touchdown, the aircraft continue not to crab and are pushed off kilter as they slow down.

Efforts must be made to get aircraft to crab into the wind at all parts of the localizer tracking. The approach of using the offsets outside a 4nm final, and the runway heading within a 4nm final should still be a decent one, they just need to crab correctly. This will also result in aircraft following the centerline exactly rather than being slightly north of the KBOS 27 centerline due to the 220@18 wind blowing them above it.

TLDR: Need more 🦀🦀🦀🦀🦀🦀🦀🦀🦀🦀🦀🦀🦀🦀🦀🦀🦀

**Comment by erikquinn** [comment]
Further support of the crabbing issue here by slowing aircraft to 5 knots upon landing. Track is about 060, wind is 240@18 (direct opposite alignment), showing the aircraft being blown that direction:

![image](https://user-images.githubusercontent.com/5103735/61556459-d7125b00-aa2f-11e9-9146-de7536514b33.png)


Making the wind calm keeps aircraft dead on centerline:

![image](https://user-images.githubusercontent.com/5103735/61556882-c4e4ec80-aa30-11e9-96df-f91ddf58698c.png)
